ash-swapchain/examples/demo.rs

```diff
@@ -0,0 +1,343 @@
+    event_loop::{ControlFlow, EventLoop},
+    window::{Window, WindowBuilder},
+};
+
+fn main() {
+    let event_loop = EventLoop::new();
+    let window = WindowBuilder::new().build(&event_loop).unwrap();
+
+    let mut app = App::new(&window);
+
+    event_loop.run(move |event, _, control_flow| {
+        *control_flow = ControlFlow::Poll;
+        match event {
+            Event::WindowEvent {
+                event: WindowEvent::CloseRequested,
+                ..
+            } => *control_flow = ControlFlow::Exit,
+            Event::WindowEvent {
+                event: WindowEvent::Resized(size),
+                ..
+            } => {
+                app.resize(vk::Extent2D {
+                    width: size.width,
+                    height: size.height,
+                });
+            }
+            Event::MainEventsCleared => {
+                app.draw();
+            }
+            _ => (),
+        }
+    });
+}
+
+pub struct App {
+    _entry: ash::Entry,
+    instance: ash::Instance,
+    surface: vk::SurfaceKHR,
+    epoch: Instant,
+
+    functions: Functions,
+    swapchain: Swapchain,
+    queue: vk::Queue,
+
+    command_pool: vk::CommandPool,
+    frames: Vec<Frame>,
+}
+
+impl App {
+    pub fn new(window: &Window) -> Self {
+        unsafe {
+            let entry = ash::Entry::new();
+            let exts = ash_window::enumerate_required_extensions(&window).unwrap();
+            let ext_ptrs = exts
+                .iter()
+                .map(|x| x.as_ptr() as *const _)
+                .collect::<Vec<_>>();
+            let instance = entry
+                .create_instance(
+                    &vk::InstanceCreateInfo::builder()
+                        .application_info(&vk::ApplicationInfo {
+                            api_version: vk::make_api_version(0, 1, 0, 0),
+                            ..Default::default()
+                        })
+                        .enabled_extension_names(&ext_ptrs),
+                    None,
+                )
+                .unwrap();
+            let surface_fn = khr::Surface::new(&entry, &instance);
+            let surface = ash_window::create_surface(&entry, &instance, &window, None).unwrap();
+
+            let (physical_device, queue_family_index) = instance
+                .enumerate_physical_devices()
+                .unwrap()
+                .into_iter()
+                .find_map(|dev| {
+                    let (family, _) = instance
+                        .get_physical_device_queue_family_properties(dev)
+                        .into_iter()
+                        .enumerate()
+                        .find(|(_index, info)| {
+                            info.queue_flags.contains(vk::QueueFlags::GRAPHICS)
+                        })?;
+                    let family = family as u32;
+                    let supported = surface_fn
+                        .get_physical_device_surface_support(dev, family, surface)
+                        .unwrap();
```

Comment:
this should be `.ok()?` instead of `.unwrap()`. Maybe `new` should also return a `Result`. There are a lot of `unwrap`s in this function.